File: built-in plugins dsp/mb-util-phase.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbUtilPhasePlugin: PluginDefinition = { id: 'mb-util-phase', slug: 'mb-util-phase', name: 'MB Phase Rotator', category: 'effect', type: 'eq' as any, version: '1.0.0', description: 'Linear phase rotation for asymmetric waveform correction', author: 'Max Booster', parameters: [{ id: 'angle', name: 'Phase Angle', type: 'float', defaultValue: 0, minValue: 0, maxValue: 360, automatable: true }, { id: 'stages', name: 'Stages', type: 'float', defaultValue: 4, minValue: 1, maxValue: 12, automatable: false }], defaultPreset: { angle: 0, stages: 4 } };

export default MbUtilPhasePlugin;

File: built-in plugins dsp/mb-util-phase.cpp
/**
 * MB Phase Rotator
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Linear phase rotation for asymmetric waveform correction
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_UTIL_PHASE_H
#define MB_UTIL_PHASE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbUtilPhase : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-util-phase";
    static constexpr const char* PLUGIN_NAME    = "MB Phase Rotator";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float angle = 0f;  // range [0, 360]
    float stages = 4f;  // range [1, 12]
    };

    MbUtilPhase() = default;
    ~MbUtilPhase() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.angle = std::clamp(params.angle, 0f, 360f);
        params.stages = std::clamp(params.stages, 1f, 12f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Phase Rotator
        return input;
    }
};

#endif // MB_UTIL_PHASE_H
